### PR TITLE
Allow ligand in water decoupling simulations

### PR DESCRIPTION
**Pull request by xiki-tempula**
Allow the BSS.FreeEnergy.Relative to handle decoupled molecule
Allow the BSS.Protocol.Config to edit the mdp file such that decoupled molecule could be written
Add new method BSS.Align.Decouple to make molecule as being decoupled
Allow BSS.mol and BSS.system to handle decoupled state properly

Add test for the above points

### CODE REVIEW

File: python/BioSimSpace/Sandpit/Exscientia/Align/_decouple.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    if not isinstance(molecule, _Molecule):
+        raise TypeError("'molecule' must be of type 'BioSimSpace._SireWrappers.Molecule'")
+
+    # Cannot decouple a perturbable molecule.
+    if molecule.isDecoupled():
+        raise _IncompatibleError("'molecule' has already been decoupled!")
+
+    if property_map0 is None and property_map1 is None:
+        property_map0 = {"charge": True, "LJ": True}
+        property_map1 = {"charge": False, "LJ": False}
```
Comment (msuruzhon) [inline python/BioSimSpace/Sandpit/Exscientia/Align/_decouple.py:54]: So the molecule is completely decoupled at `fep-lambda=1`, is that right?

Comment (xiki-tempula) [reply python/BioSimSpace/Sandpit/Exscientia/Align/_decouple.py:54]: If `property_map1 = {"charge": False, "LJ": False}` then yes.

File: python/BioSimSpace/Sandpit/Exscientia/Align/_decouple.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    # Edit the molecule
+    mol_edit = mol_sire.edit()
+
+    # Set the start and the end state
+    mol_edit.setProperty("charge0", _SireBase.wrap(property_map0.get("charge",True)))
+    mol_edit.setProperty("charge1", _SireBase.wrap(property_map1.get("charge", True)))
+    mol_edit.setProperty("LJ0", _SireBase.wrap(property_map0.get("LJ", True)))
+    mol_edit.setProperty("LJ1", _SireBase.wrap(property_map1.get("LJ", True)))
+
+    mol_edit.setProperty("intramol", _SireBase.wrap(intramol))
```
Comment (msuruzhon) [inline python/BioSimSpace/Sandpit/Exscientia/Align/_decouple.py:86]: Would it be helpful to call it something more transparent, like e.g. `is_annihilated`?

Comment (xiki-tempula) [reply python/BioSimSpace/Sandpit/Exscientia/Align/_decouple.py:86]: @msuruzhon So I use only `isDecoupled` to check if the molecule has been Decoupled or Annihilated.
Then I don't need to check `mol.isDecoupled() or mol.isAnnihilated()` and can just check `mol.isDecoupled()`

File: python/BioSimSpace/Sandpit/Exscientia/FreeEnergy/_relative.py [hunk 3]
```diff
@@ -216,7 +216,7 @@ def __init__(self, system, protocol=None, work_dir=None, engine=None,
                     raise _MissingSoftwareError("Cannot use GROMACS engine as GROMACS is not installed!")
 
                 # The system must have a perturbable molecule.
-                if system.nPerturbableMolecules() == 0:
+                if system.nPerturbableMolecules() == 0 and system.nDecoupledMolecules() == 0:
```
Comment (msuruzhon) [inline python/BioSimSpace/Sandpit/Exscientia/FreeEnergy/_relative.py:219]: We wouldn't expect a decoupled molecule in a relative free energy calculation? Would it be worth adding this in once you have ABFE support?

Comment (xiki-tempula) [reply python/BioSimSpace/Sandpit/Exscientia/FreeEnergy/_relative.py:219]: @msuruzhon For Gromacs charge-perturbed ABFE, the charge molecule is decoupled and the ion is perturbed alchemically to account for the charge difference, so we need both.

File: test/Sandpit/Exscientia/Align/test_decouple.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    mol = BSS.Parameters.openff_unconstrained_2_0_0(
+        "c1ccccc1").getMolecule()
+    return mol
+
+def test_sanity(mol):
+    assert mol.isDecoupled() == False
+
+def test_set_property_map(mol):
+    '''Test if the charge and LJ are set properly'''
+    for values in itertools.product(*[(True, False) for _ in range(4)]):
```
Comment (msuruzhon) [inline test/Sandpit/Exscientia/Align/test_decouple.py:20]: Would it be better to move this loop out of the function as a `pytest.mark.parametrize`?

**Review by msuruzhon — APPROVED** [review]
LGTM apart from some minor suggestions

**Pull request merged**